sklearn/utils/multiclass.py

```diff
@@ -414,7 +414,11 @@ def _raise_or_return():
     if issparse(first_row_or_val):
         first_row_or_val = first_row_or_val.data
     classes = cached_unique(y)
-    if y.shape[0] > 20 and classes.shape[0] > round(0.5 * y.shape[0]):
+    if (
+        not suppress_warning
+        and y.shape[0] > 20
+        and classes.shape[0] > round(0.5 * y.shape[0])
+    ):
```

Comment:
Since the spurious warning is only triggered because we call `type_of_target` on classes, I'd prefer to add an extra condition to avoid it rather than adding a new parameter to a public function.

```suggestion
    if y.shape[0] > 20 and y.shape[0] > classes.shape[0] > round(0.5 * y.shape[0]):
```

With this suggestion, there's an edge case where a `y` contains only unique values but it's kind of useless for a classification task so a lot of problems would have happen before that, so I wouldn't worry about it.

Reply:
This test creates a `y` containing only unique values. Let's change the generated `y` by
`y = np.repeat(np.arange(20), 2)`

Reply:
Oh right because now it's less than 50%. Let's do `y = np.concat((np.arange(20), [0]))` to have > 20 samples and > 50% unique classes